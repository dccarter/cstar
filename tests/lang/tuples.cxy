// tuple declaration
type Tup1 = (bool, wchar)
type Tup2 = (wchar, (f32, string, Tup1))

func main(args: [string]) {
    // tuple literals
    // NOTE (x) is not a tuple literal
    (true, 10);
    ('a', true, 10, "Hello");
    ('a', ("World", 32, (true, 5.996)));

    // Tuple type checking
    var x: (bool, i32) = (true, 10);
    x = (false, 20)
    var y : Tup2 = ('a', (0.6, "Hello", (true, 'A')));
}