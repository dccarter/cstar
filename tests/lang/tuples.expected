{
 "tag": "Program",
 "decls": [{
   "tag": "TypeDecl",
   "flags": "(TopLevelDecl)",
   "name": "Tup1",
   "aliased": {
    "tag": "TupleType",
    "flags": "(TypeAst)",
    "members": [{
      "tag": "PrimitiveType",
      "flags": "(TypeAst)",
      "prtId": 0
     }, {
      "tag": "PrimitiveType",
      "flags": "(TypeAst)",
      "prtId": 1
     }]
   }
  }, {
   "tag": "TypeDecl",
   "flags": "(TopLevelDecl)",
   "name": "Tup2",
   "aliased": {
    "tag": "TupleType",
    "flags": "(TypeAst)",
    "members": [{
      "tag": "PrimitiveType",
      "flags": "(TypeAst)",
      "prtId": 1
     }, {
      "tag": "TupleType",
      "flags": "(TypeAst)",
      "members": [{
        "tag": "PrimitiveType",
        "flags": "(TypeAst)",
        "prtId": 11
       }, {
        "tag": "StringType",
        "flags": "(TypeAst)"
       }, {
        "tag": "Path",
        "flags": "(TypeAst)",
        "elements": [{
          "tag": "PathElem",
          "name": "Tup1",
          "index": 0
         }],
        "isType": true
       }]
     }]
   }
  }, {
   "tag": "FuncDecl",
   "flags": "(TopLevelDecl)",
   "overload": 51,
   "index": 0,
   "name": "swap",
   "params": [{
     "tag": "FuncParam",
     "name": "tup",
     "index": 0,
     "paramType": {
      "tag": "TupleType",
      "flags": "(TypeAst)",
      "members": [{
        "tag": "StringType",
        "flags": "(TypeAst)"
       }, {
        "tag": "PrimitiveType",
        "flags": "(TypeAst)",
        "prtId": 9
       }]
     }
    }],
   "ret": {
    "tag": "TupleType",
    "flags": "(TypeAst)",
    "members": [{
      "tag": "PrimitiveType",
      "flags": "(TypeAst)",
      "prtId": 9
     }, {
      "tag": "StringType",
      "flags": "(TypeAst)"
     }]
   },
   "body": {
    "tag": "TupleExpr",
    "members": [{
      "tag": "MemberExpr",
      "target": {
       "tag": "Path",
       "elements": [{
         "tag": "PathElem",
         "name": "tup",
         "index": 0
        }],
       "isType": false
      },
      "member": {
       "tag": "IntegerLit",
       "value": 1
      }
     }, {
      "tag": "MemberExpr",
      "target": {
       "tag": "Path",
       "elements": [{
         "tag": "PathElem",
         "name": "tup",
         "index": 0
        }],
       "isType": false
      },
      "member": {
       "tag": "IntegerLit",
       "value": 0
      }
     }]
   }
  }, {
   "tag": "FuncDecl",
   "flags": "(TopLevelDecl)",
   "overload": 51,
   "index": 0,
   "name": "main",
   "params": [{
     "tag": "FuncParam",
     "name": "args",
     "index": 0,
     "paramType": {
      "tag": "ArrayType",
      "flags": "(TypeAst)",
      "element": {
       "tag": "StringType",
       "flags": "(TypeAst)"
      }
     }
    }],
   "body": {
    "tag": "BlockStmt",
    "stmts": [{
      "tag": "ExprStmt",
      "expr": {
       "tag": "TupleExpr",
       "members": [{
         "tag": "BoolLit",
         "value": 1
        }, {
         "tag": "IntegerLit",
         "value": 10
        }]
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "TupleExpr",
       "members": [{
         "tag": "CharLit",
         "value": 97
        }, {
         "tag": "BoolLit",
         "value": 1
        }, {
         "tag": "IntegerLit",
         "value": 10
        }, {
         "tag": "StringLit",
         "value": "Hello"
        }]
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "TupleExpr",
       "members": [{
         "tag": "CharLit",
         "value": 97
        }, {
         "tag": "TupleExpr",
         "members": [{
           "tag": "StringLit",
           "value": "World"
          }, {
           "tag": "IntegerLit",
           "value": 32
          }, {
           "tag": "TupleExpr",
           "members": [{
             "tag": "BoolLit",
             "value": 1
            }, {
             "tag": "FloatLit",
             "value": 5.996000
            }]
          }]
        }]
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "x"
       }],
      "varType": {
       "tag": "TupleType",
       "flags": "(TypeAst)",
       "members": [{
         "tag": "PrimitiveType",
         "flags": "(TypeAst)",
         "prtId": 0
        }, {
         "tag": "PrimitiveType",
         "flags": "(TypeAst)",
         "prtId": 9
        }]
      },
      "init": {
       "tag": "TupleExpr",
       "members": [{
         "tag": "BoolLit",
         "value": 1
        }, {
         "tag": "IntegerLit",
         "value": 10
        }]
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "AssignExpr",
       "op": 34,
       "lhs": {
        "tag": "Path",
        "elements": [{
          "tag": "PathElem",
          "name": "x",
          "index": 0
         }],
        "isType": false
       },
       "rhs": {
        "tag": "Path",
        "elements": [{
          "tag": "PathElem",
          "name": "x",
          "index": 0
         }],
        "isType": false
       }
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "y"
       }],
      "varType": {
       "tag": "Path",
       "flags": "(TypeAst)",
       "elements": [{
         "tag": "PathElem",
         "name": "Tup2",
         "index": 0
        }],
       "isType": true
      },
      "init": {
       "tag": "TupleExpr",
       "members": [{
         "tag": "CharLit",
         "value": 97
        }, {
         "tag": "TupleExpr",
         "members": [{
           "tag": "FloatLit",
           "value": 0.600000
          }, {
           "tag": "StringLit",
           "value": "Hello"
          }, {
           "tag": "TupleExpr",
           "members": [{
             "tag": "BoolLit",
             "value": 1
            }, {
             "tag": "CharLit",
             "value": 65
            }]
          }]
        }]
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "AssignExpr",
       "op": 34,
       "lhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "Path",
         "elements": [{
           "tag": "PathElem",
           "name": "x",
           "index": 0
          }],
         "isType": false
        },
        "member": {
         "tag": "IntegerLit",
         "value": 0
        }
       },
       "rhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "Path",
         "elements": [{
           "tag": "PathElem",
           "name": "x",
           "index": 0
          }],
         "isType": false
        },
        "member": {
         "tag": "IntegerLit",
         "value": 0
        }
       }
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "AssignExpr",
       "op": 34,
       "lhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "Path",
         "elements": [{
           "tag": "PathElem",
           "name": "x",
           "index": 0
          }],
         "isType": false
        },
        "member": {
         "tag": "IntegerLit",
         "value": 1
        }
       },
       "rhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "Path",
         "elements": [{
           "tag": "PathElem",
           "name": "x",
           "index": 0
          }],
         "isType": false
        },
        "member": {
         "tag": "IntegerLit",
         "value": 1
        }
       }
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "AssignExpr",
       "op": 34,
       "lhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "MemberExpr",
         "target": {
          "tag": "MemberExpr",
          "target": {
           "tag": "Path",
           "elements": [{
             "tag": "PathElem",
             "name": "y",
             "index": 0
            }],
           "isType": false
          },
          "member": {
           "tag": "IntegerLit",
           "value": 1
          }
         },
         "member": {
          "tag": "IntegerLit",
          "value": 2
         }
        },
        "member": {
         "tag": "IntegerLit",
         "value": 1
        }
       },
       "rhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "MemberExpr",
         "target": {
          "tag": "MemberExpr",
          "target": {
           "tag": "Path",
           "elements": [{
             "tag": "PathElem",
             "name": "y",
             "index": 0
            }],
           "isType": false
          },
          "member": {
           "tag": "IntegerLit",
           "value": 1
          }
         },
         "member": {
          "tag": "IntegerLit",
          "value": 2
         }
        },
        "member": {
         "tag": "IntegerLit",
         "value": 1
        }
       }
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "AssignExpr",
       "op": 34,
       "lhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "MemberExpr",
         "target": {
          "tag": "Path",
          "elements": [{
            "tag": "PathElem",
            "name": "y",
            "index": 0
           }],
          "isType": false
         },
         "member": {
          "tag": "IntegerLit",
          "value": 1
         }
        },
        "member": {
         "tag": "IntegerLit",
         "value": 0
        }
       },
       "rhs": {
        "tag": "MemberExpr",
        "target": {
         "tag": "MemberExpr",
         "target": {
          "tag": "Path",
          "elements": [{
            "tag": "PathElem",
            "name": "y",
            "index": 0
           }],
          "isType": false
         },
         "member": {
          "tag": "IntegerLit",
          "value": 1
         }
        },
        "member": {
         "tag": "IntegerLit",
         "value": 0
        }
       }
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "z"
       }],
      "varType": {
       "tag": "TupleType",
       "flags": "(TypeAst)",
       "members": [{
         "tag": "PrimitiveType",
         "flags": "(TypeAst)",
         "prtId": 9
        }, {
         "tag": "StringType",
         "flags": "(TypeAst)"
        }]
      },
      "init": {
       "tag": "CallExpr",
       "callee": {
        "tag": "Path",
        "elements": [{
          "tag": "PathElem",
          "name": "swap",
          "index": 0
         }],
        "isType": false
       },
       "args": [{
         "tag": "TupleExpr",
         "members": [{
           "tag": "StringLit",
           "value": "Hello"
          }, {
           "tag": "IntegerLit",
           "value": 10
          }]
        }],
       "overload": 0
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "f"
       }],
      "varType": {
       "tag": "PointerType",
       "flags": "(TypeAst)",
       "pointed": {
        "tag": "TupleType",
        "flags": "(TypeAst)",
        "members": [{
          "tag": "PrimitiveType",
          "flags": "(TypeAst)",
          "prtId": 9
         }, {
          "tag": "StringType",
          "flags": "(TypeAst)"
         }]
       }
      },
      "init": {
       "tag": "AddressOf",
       "__fallback": null
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "k"
       }],
      "varType": {
       "tag": "TupleType",
       "flags": "(TypeAst)",
       "members": [{
         "tag": "StringType",
         "flags": "(TypeAst)"
        }, {
         "tag": "PrimitiveType",
         "flags": "(TypeAst)",
         "prtId": 9
        }]
      },
      "init": {
       "tag": "TupleExpr",
       "members": [{
         "tag": "MemberExpr",
         "target": {
          "tag": "Path",
          "elements": [{
            "tag": "PathElem",
            "name": "f",
            "index": 0
           }],
          "isType": false
         },
         "member": {
          "tag": "IntegerLit",
          "value": 1
         }
        }, {
         "tag": "MemberExpr",
         "target": {
          "tag": "Path",
          "elements": [{
            "tag": "PathElem",
            "name": "f",
            "index": 0
           }],
          "isType": false
         },
         "member": {
          "tag": "IntegerLit",
          "value": 0
         }
        }]
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "g"
       }],
      "init": {
       "tag": "UnaryExpr",
       "op": 23,
       "isPrefix": true,
       "lhs": {
        "tag": "Path",
        "elements": [{
          "tag": "PathElem",
          "name": "f",
          "index": 0
         }],
        "isType": false
       }
      }
     }]
   }
  }]
}
